### PR TITLE
[feature][generator] Added "service" geometry offset flag.

### PR DESCRIPTION
**Pull request by vng**
Follow up https://github.com/organicmaps/organicmaps/pull/5540

### CODE REVIEW

File: indexer/feature.hpp [hunk 1]
```diff
@@ -17,7 +17,10 @@ namespace feature
 {
 class SharedLoadInfo;
 struct NameParamsOut; // Include feature_utils.hpp when using
-}
+
+// 1 is a "service" flag value. We are sure that there is no real geometry offset == 1.
```
Comment (pastk) [inline indexer/feature.hpp:21]:
```suggestion
// 1 is a "fallback" flag value, which means geometry should be loaded from the next offset of the more detailed geom level.
```

**Review by pastk — APPROVED** [review]
Looks great! (only the comment should be fixed)

Its better than the https://github.com/organicmaps/organicmaps/pull/5540#issuecomment-1646820064 solution, because it uses a shortest 1-byte marker offset instead of a longer real offset.

File: indexer/feature.hpp [hunk 2]
```diff
@@ -17,7 +17,10 @@ namespace feature
...
+uint32_t constexpr kHasGeoOffsetFlag = 1;
```
Comment (biodranik) [inline indexer/feature.hpp:22]: Is there a better name for this constant?

**Review by biodranik — COMMENTED** [review]
1. What are the real data savings for the same maps data?
2. Will older maps work?

File: indexer/feature.hpp (hunk 2 above)
Comment (pastk) [reply indexer/feature.hpp:22]: `kGeomOffsetFallback`?

**Comment by pastk** [comment]
> 1. What are the real data savings for the same maps data?
> 

A tiny bit less savings than #5569.

>     2. Will older maps work?

Yes.

File: generator/geometry_holder.hpp [hunk 3]
```diff
@@ -96,6 +93,12 @@ class GeometryHolder
         WriteOuterPoints(points, scaleIndex);
         m_ptsPrevCount = points.size();
       }
+      else
+      {
+        CHECK(m_buffer.m_ptsMask != 0, ("Some valid geometry should present"));
```
Comment (pastk) [inline generator/geometry_holder.hpp:98]:
```suggestion
        CHECK(m_buffer.m_ptsMask != 0, ("Some valid geometry should be present already"));
```

File: generator/geometry_holder.hpp [hunk 4]
```diff
@@ -152,14 +155,25 @@ class GeometryHolder
...
-      m_trgPrevCount = trgPointsCount;
+      if (WriteOuterTriangles(polys, scaleIndex))
+      {
+        // Assign only if geometry is valid (correctly tesselated and saved).
+        m_trgPrevCount = trgPointsCount;
+      }
+    }
+    else
+    {
+      CHECK(m_buffer.m_trgMask != 0, ("Some valid geometry should present"));
```
Comment (pastk) [inline generator/geometry_holder.hpp:174]:
```suggestion
      CHECK(m_buffer.m_trgMask != 0, ("Some valid geometry should be present already"));
```

**Review by pastk — APPROVED** [review]

File: indexer/feature.cpp [hunk 5]
```diff
@@ -89,15 +94,16 @@ int GetScaleIndex(SharedLoadInfo const & loadInfo, int scale,
...
+    // If there is no geometry for the requested scale (kHasGeoOffsetFlag) fallback to the next more detailed one.
+    while (ind < count && (scale > loadInfo.GetScale(ind) || offsets[ind] == kGeomOffsetFallback))
       ++ind;
+
     // Some WorldCoasts features have idx == 0 geometry only and its possible
     // other features to be visible on e.g. idx == 1 only,
     // but then they shouldn't be attempted to be drawn using other geom scales.
     ASSERT_LESS(ind, count, ("No suitable geometry scale range in the map file."));
-    return (ind < count ? ind : -1);
+    return (offsets[ind] != kInvalidOffset ? ind : -1);
```
Comment (biodranik) [inline indexer/feature.cpp:106]: Is fallback offset ok here?

**Review by biodranik — APPROVED** [review]

File: indexer/feature.cpp (hunk 5 above)
Comment (vng) [reply indexer/feature.cpp:106]: We won't get here with kGeomOffsetFallback

**Comment by biodranik** [comment]
Right. Only new app + old mwm is possible.

**Comment by Misalf-git** [comment]
> But this old app + new mwm situation should never happen, right?
Except when manually downloading maps from [https://cdn.organicmaps.app/](Map data mirrors for Organic Maps), for example for using other ways of moving them to a device without internet connection.

**Comment by biodranik** [comment]
We do not support old app + new mwms and will never support this scenario. 

Those who install mwms manually should always install the latest app version.

**Pull request merged**